### PR TITLE
[Orders] After changing tariff the price of paid order is not changing 

### PR DESCRIPTION
**Pull request by ospodaryk**
# GreenCityUBS PR
[[Orders] After changing tariff the price of paid order is changing #6051
](https://github.com/ita-social-projects/GreenCity/issues/6051)

[[Orders] After changing tariff and increasing in the number of ordered products for a paid order the price of paid order is changing according to the new price #6070](https://github.com/ita-social-projects/GreenCity/issues/6070)

[[Orders] After changing tariff and decreasing in the number of ordered products for a paid order the price of paid order changes according to the new price #6071](https://github.com/ita-social-projects/GreenCity/issues/6071)

[[Change request to #2196] Changing cost of service #6038
](https://github.com/ita-social-projects/GreenCity/issues/6038)

## Summary Of Changes :fire:
Now when we change tariff it doesn't affect  price for Paid orders.

## Added
Added new entity OrderBag which is responsible for setting static bag prices per order.
Accordingly in order is a new field OrderBag which gets a list of OrderBags to get data from that.

## Changed
Method where we get a price for setting total sum in order.

# PR Checklist Forms

- [ ] Code is up-to-date with the `dev` branch.
- [x] You've successfully built and run the tests locally.
- [x] There are new or updated unit tests validating the change.
- [x] JIRA/ Github Issue number & title in PR title (ISSUE-XXXX: Ticket title)
- [x] This template filled (above this section).
- [x] Sonar's report does not contain bugs, vulnerabilities, security issues, code smells ar duplication
- [x] `NEED_REVIEW` and `READY_FOR_REVIEW` labels are added.
- [x] All files reviewed before sending to reviewers

### CODE REVIEW

File: dao/src/main/java/greencity/entity/order/OrderBag.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    private Order order;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "bag_id")
+    private Bag bag;
+
+    @Column(nullable = false)
+    private Integer amount;
+
+    @Column(name = "confirmed_quantity")
```
Comment (VadymMakitra) [inline dao/src/main/java/greencity/entity/order/OrderBag.java:48]: the name is not required here, Hibernate can create a mapping automatically.
eg: dateOfOrder = date_of_order; confirmedQuantity = confirmed_quantity;

File: dao/src/main/java/greencity/entity/order/OrderBag.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    @Column(nullable = false)
+    private Long price;
+
+    @Size(min = 1, max = 30)
+    @Column(nullable = false)
+    private String name;
+
+    @Size(min = 1, max = 30)
+    @Column(nullable = false)
+    private String nameEng;
```
Comment (VadymMakitra) [inline dao/src/main/java/greencity/entity/order/OrderBag.java:68]: @NotBlank

File: dao/src/main/resources/db/changelog/logs/ch-update-order-bag-mapping-table-Spodaryk.xml [hunk 3]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.4.xsd">
+    <changeSet id="Spodaryk-4" author="Oksana Spodaryk">
+        <dropPrimaryKey  tableName= "order_bag_mapping" constraintName="order_bag_mapping_pkey"/>
```
Comment (VadymMakitra) [inline dao/src/main/resources/db/changelog/logs/ch-update-order-bag-mapping-table-Spodaryk.xml]: Could you please add preCondition: https://docs.liquibase.com/concepts/changelogs/preconditions.html

File: dao/src/main/resources/db/changelog/logs/ch-update-order-bag-mapping-table-Spodaryk.xml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+        <addForeignKeyConstraint baseTableName="order_bag_mapping"
+                                 baseColumnNames="bag_id"
+                                 constraintName="fk_bag_to_order"
+                                 referencedTableName="bag"
+                                 referencedColumnNames="id"/>
+    </changeSet>
+
+    <changeSet id="Spodaryk-5" author="Oksana Spodaryk">
+        <sql>update order_bag_mapping set
```
Comment (VadymMakitra) [inline dao/src/main/resources/db/changelog/logs/ch-update-order-bag-mapping-table-Spodaryk.xml]: Please move this to the SQL file, and use the created file: https://docs.liquibase.com/change-types/sql-file.html. Also, you can add here a comment to describe why this change was be done (eg. Ticket number and what done)

File: service-api/src/main/java/greencity/constant/ErrorMessage.java [hunk 5]
```diff
@@ -163,6 +163,7 @@ public final class ErrorMessage {
     public static final String INVALID_URL = "Invalid URL: ";
     public static final String LOCATION_CAN_NOT_BE_DELETED =
         "Such location cannot be deleted as it is linked to the tariff";
+    public static final String ORDER_NOT_FOUND = "Order not found with id: ";
```
Comment (VadymMakitra) [inline service-api/src/main/java/greencity/constant/ErrorMessage.java]:
```suggestion
    public static final String ORDER_NOT_FOUND = "Order with the selected id does not exist";
```

Comment (VadymMakitra) [reply service-api/src/main/java/greencity/constant/ErrorMessage.java]: If you want to add ID in the message you can use String.format()

File: dao/src/main/java/greencity/entity/order/OrderBag.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    private Integer confirmedQuantity;
+
+    @Column(name = "exported_quantity")
+    private Integer exportedQuantity;
+
+    @Column(nullable = false)
+    private Integer capacity;
+
+    @Column(nullable = false)
+    private Long price;
```
Comment (VadymMakitra) [inline dao/src/main/java/greencity/entity/order/OrderBag.java:58]:
```suggestion
    private BigDecimal price;
```

File: service/src/main/java/greencity/service/ubs/OrderBagService.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+@Service
+@AllArgsConstructor
+@Slf4j
+@Data
+public class OrderBagService {
+    private OrderBagRepository orderBagRepository;
+
+    private Long getActualPrice(List<OrderBag> orderBags, Integer id) {
+        return orderBags.stream()
+            .filter(ob -> ob.getBag().getId().equals(id))
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/OrderBagService.java:30]: Can you filter this in SQL, instead of filtering in the JAVA code?

File: service/src/main/java/greencity/service/ubs/OrderBagService.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+     * @return A list of Bag instances associated with the provided OrderBag
+     *         instances.
+     */
+    public List<Bag> findAllBagsInOrderBagsList(List<OrderBag> orderBags) {
+        return orderBags.stream()
+            .map(OrderBag::getBag)
+            .map(b -> {
+                b.setFullPrice(getActualPrice(orderBags, b.getId()));
+                return b;
+            })
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/OrderBagService.java]:
```suggestion
            .peek(bag -> {
                bag.setFullPrice(getActualPrice(orderBags, b.getId()));
            })
```

File: service/src/main/java/greencity/service/ubs/OrderBagService.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+     *         amount as the value. If any OrderBag in the input list lacks all
+     *         three attributes (exportedQuantity, confirmedQuantity, and amount),
+     *         the corresponding entry will not be included in the result map.
+     * @throws NullPointerException if 'bagsForOrder' is null.
+     */
+    public Map<Integer, Integer> getActualBagsAmountForOrder(List<OrderBag> bagsForOrder) {
+        if (bagsForOrder.stream().allMatch(it -> it.getExportedQuantity() != null)) {
+            return bagsForOrder.stream()
+                .collect(Collectors.toMap(it -> it.getBag().getId(), OrderBag::getExportedQuantity));
+        }
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/OrderBagService.java:85]:
```suggestion
        if (bagsForOrder.stream().allMatch(orderBag -> orderBag.getExportedQuantity() != null)) {
            return bagsForOrder.stream()
                .collect(Collectors.toMap(orderBag -> orderBag.getBag().getId(), OrderBag::getExportedQuantity));
        }
```

File: service/src/main/java/greencity/service/ubs/OrderBagService.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        if (bagsForOrder.stream().allMatch(it -> it.getConfirmedQuantity() != null)) {
+            return bagsForOrder.stream()
+                .collect(Collectors.toMap(it -> it.getBag().getId(), OrderBag::getConfirmedQuantity));
+        }
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/OrderBagService.java:89]:
```suggestion
        if (bagsForOrder.stream().allMatch(orderBag -> orderBag.getConfirmedQuantity() != null)) {
            return bagsForOrder.stream()
                .collect(Collectors.toMap(orderBag -> orderBag.getBag().getId(), OrderBag::getConfirmedQuantity));
        }
```

File: service/src/main/java/greencity/service/ubs/OrderBagService.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        if (bagsForOrder.stream().allMatch(it -> it.getAmount() != null)) {
+            return bagsForOrder.stream()
+                .collect(Collectors.toMap(it -> it.getBag().getId(), OrderBag::getAmount));
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/OrderBagService.java:92]:
```suggestion
        if (bagsForOrder.stream().allMatch(orderBag -> orderBag.getAmount() != null)) {
            return bagsForOrder.stream()
                .collect(Collectors.toMap(orderBag -> orderBag.getBag().getId(), OrderBag::getAmount));
```

File: service/src/main/java/greencity/service/ubs/OrderBagService.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    /**
+     * method helps to delete bag from order.
+     *
+     * @param orderBag {@link OrderBag}
+     * @author Oksana Spodaryk
+     */
+    public void removeBagFromOrder(Order order, OrderBag orderBag) {
+        List<OrderBag> modifiableList = new ArrayList<>(order.getOrderBags());
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/OrderBagService.java:104]: Could you please explain why you create new list, as i can see from the order entity should return mutable collection.

File: service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java [hunk 13]
```diff
@@ -142,13 +151,34 @@ public List<GetTariffServiceDto> getTariffService(long tariffId) {
...
+        Integer totalBagsAmount = amount.values().stream().reduce(0, Integer::sum);
+        if (amount.get(bagId).equals(0) || order.getOrderPaymentStatus() == OrderPaymentStatus.UNPAID) {
+            if (totalBagsAmount.equals(amount.get(bagId))) {
+                order.setOrderBags(new ArrayList<>());
+                orderRepository.delete(order);
+                return;
+            }
+            order.getOrderBags().stream().filter(it -> it.getBag().getId().equals(bagId))
+                .findFirst()
+                .ifPresent(obj -> orderBagService.removeBagFromOrder(order, obj));
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java]:
```suggestion
            order.getOrderBags().stream().filter(orderBag -> orderBag.getBag().getId().equals(bagId))
                .findFirst()
                .ifPresent(orderBag -> orderBagService.removeBagFromOrder(order, orderBag));
```

File: service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java [hunk 14]
```diff
@@ -159,9 +189,39 @@ public void deleteTariffService(Integer bagId) {
...
+        Bag bag = tryToFindBagById(bagId);
         Employee employee = tryToFindEmployeeByUuid(employeeUuid);
+        updateTariffService(dto, bag);
+        bag.setEditedBy(employee);
+
+        orderBagRepository.updateAllByBagIdForUnpaidOrders(
+            bagId, bag.getCapacity(), bag.getFullPrice(), bag.getName(), bag.getNameEng());
+
+        List<Order> orders = orderRepository.findAllUnpaidOrdersByBagId(bagId);
+        if (!orders.isEmpty()) {
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java]:
```suggestion
        if (CollectionUtils.isNotEmpty(orders)) {
```

File: service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java [hunk 15]
```diff
@@ -159,9 +189,39 @@ public void deleteTariffService(Integer bagId) {
...
+            orders.forEach(it -> updateOrderSumToPay(it, bag));
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java:204]:
```suggestion
            orders.forEach(order -> updateOrderSumToPay(order, bag));
```

File: service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java [hunk 16]
```diff
@@ -159,9 +189,39 @@ public void deleteTariffService(Integer bagId) {
...
+            orderRepository.saveAll(orders);
+        }
+        return modelMapper.map(bagRepository.save(bag), GetTariffServiceDto.class);
+    }
+
+    private void updateOrderSumToPay(Order order, Bag bag) {
+        Map<Integer, Integer> amount = orderBagService.getActualBagsAmountForOrder(order.getOrderBags());
+        Long sumToPayInCoins = order.getOrderBags().stream()
+            .map(it -> amount.get(it.getBag().getId()) * getActualBagPrice(it, bag))
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java]:
```suggestion
            .map(orderBag -> amount.get(orderBag.getBag().getId()) * getActualBagPrice(orderBag, bag))
```

File: service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java [hunk 17]
```diff
@@ -159,9 +189,39 @@ public void deleteTariffService(Integer bagId) {
...
+            .reduce(0L, Long::sum);
+        order.setSumTotalAmountWithoutDiscounts(sumToPayInCoins);
+    }
+
+    private Long getActualBagPrice(OrderBag orderBag, Bag bag) {
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java]:
```suggestion
    private Long getBagPrice(OrderBag orderBag, Bag bag) {
```

File: service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java [hunk 18]
```diff
@@ -215,14 +273,16 @@ private Service createService(Long tariffId, ServiceDto dto, String employeeUuid
...
             .map(it -> modelMapper.map(it, GetServiceDto.class))
-            .orElse(null);
+            .orElseGet(() -> null);
     }
 
     @Override
     public void deleteService(long id) {
-        serviceRepository.delete(tryToFindServiceById(id));
+        Service service = tryToFindServiceById(id);
+        serviceRepository.delete(service);
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java:285]:
```suggestion
        serviceRepository.deleteById(id);
```

Comment (VadymMakitra) [reply service/src/main/java/greencity/service/ubs/SuperAdminServiceImpl.java:285]: I think we can remove the service in this way. It doesn't matter if this repo exists or no, we remove it anyway.

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 19]
```diff
@@ -457,7 +458,7 @@ private List<Integer> getBagIds(List<BagDto> dto) {
             .collect(Collectors.toList());
     }
 
-    private Bag tryToGetBagById(Integer id) {
+    private Bag tryToGetActiveBagById(Integer id) {
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]:
```suggestion
    private Bag findBagById(Integer id) {
```

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 20]
```diff
@@ -885,31 +886,17 @@ private AddressInfoDto addressInfoDtoBuilder(Order order) {
     }
 
     private List<BagForUserDto> bagForUserDtosBuilder(Order order) {
-        Map<Integer, Integer> actualBagAmounts = getActualBagAmountsForOrder(order);
-        List<Bag> bagsForOrder = bagRepository.findBagsByOrderId(order.getId());
+        List<OrderBag> bagsForOrder = order.getOrderBags();
+        Map<Integer, Integer> actualBagsAmount = orderBagService.getActualBagsAmountForOrder(bagsForOrder);
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java:890]:
```suggestion
        Map<Integer, Integer> actualBagsAmount = orderBagService.getActualBagsAmountForOrder(order.getOrderBags());
```

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 21]
```diff
@@ -1038,15 +1025,16 @@ private UBSuser updateRecipientDataInOrder(UBSuser ubsUser, UbsCustomersDtoUpdat
     }
 
     private Order formAndSaveOrder(Order order, Set<Certificate> orderCertificates,
-        Map<Integer, Integer> amountOfBagsOrderedMap, UBSuser userData,
+        List<OrderBag> bagsOrdered, UBSuser userData,
         User currentUser, long sumToPayInCoins) {
         order.setOrderStatus(OrderStatus.FORMED);
         order.setCertificates(orderCertificates);
-        order.setAmountOfBagsOrdered(amountOfBagsOrderedMap);
+        bagsOrdered.forEach(it -> it.setOrder(order));
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]:
```suggestion
        bagsOrdered.forEach(orderBag -> orderBag.setOrder(order));
```

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 22]
```diff
@@ -1202,35 +1190,46 @@ private void checkAmountOfBagsIfCourierLimitByAmountOfBag(TariffsInfo courierLoc
...
-        long sumToPayInCoins = 0L;
-
-        for (Map.Entry<Integer, Integer> temp : getOrderBagsAndQuantity.entrySet()) {
-            Integer amount = getOrderBagsAndQuantity.get(temp.getKey());
-            Bag bag = bagRepository.findById(temp.getKey())
-                .orElseThrow(() -> new NotFoundException(BAG_NOT_FOUND + temp.getKey()));
-            sumToPayInCoins += bag.getFullPrice() * amount;
-        }
-        return sumToPayInCoins;
+    private long calculateOrderSumWithoutDiscounts(List<OrderBag> getOrderBagsAndQuantity) {
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java:1193]:
```suggestion
    private BigDecimal calculateOrderSumWithoutDiscounts(List<OrderBag> getOrderBagsAndQuantity) {
```

Comment (VadymMakitra) [reply service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java:1193]: I think that we should include coins in the calculation

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 23]
```diff
@@ -1202,35 +1190,46 @@ private void checkAmountOfBagsIfCourierLimitByAmountOfBag(TariffsInfo courierLoc
...
+        return getOrderBagsAndQuantity.stream()
+            .map(it -> it.getPrice() * it.getAmount())
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]:
```suggestion
            .map(orderBag -> orderBag.getPrice() * orderBag.getAmount())
```

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 24]
```diff
@@ -1202,35 +1190,46 @@ private void checkAmountOfBagsIfCourierLimitByAmountOfBag(TariffsInfo courierLoc
...
+            OrderBag orderBag = createOrderBag(bag);
+            orderBag.setAmount(temp.getAmount());
+            orderBagList.add(orderBag);
+        }
+        List<Integer> orderedBagsIds = bags.stream().map(BagDto::getId).collect(toList());
+        List<OrderBag> notOrderedBags = tariffsInfo.getBags().stream()
+            .filter(it -> !orderedBagsIds.contains(it.getId()))
+            .map(this::createOrderBag).collect(toList());
+        notOrderedBags.forEach(it -> it.setAmount(0));
+        orderBagList.addAll(notOrderedBags);
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]:
```suggestion
        List<Integer> orderedBagsIds = bags.stream().map(BagDto::getId).collect(toList());
        List<OrderBag> notOrderedBags = tariffsInfo.getBags().stream()
            .filter(orderBag -> !orderedBagsIds.contains(orderBag.getId()))
            .map(this::createOrderBag).peek(orderBag -> orderBag.setAmount(0)).collect(toList());
        orderBagList.addAll(notOrderedBags);
```

**Review by VadymMakitra — COMMENTED** [review]
Please fix comments

File: dao/src/main/java/greencity/entity/order/OrderBag.java (hunk 6 above)
Comment (ospodaryk) [reply dao/src/main/java/greencity/entity/order/OrderBag.java:58]: Throughout the program we use Long for price. Therefore, I am not sure how appropriate it is to change it

File: service/src/main/java/greencity/service/ubs/OrderBagService.java (hunk 8 above)
Comment (ospodaryk) [reply service/src/main/java/greencity/service/ubs/OrderBagService.java]: when I used peek sonar it gave me codesmell

File: service/src/main/java/greencity/service/ubs/OrderBagService.java (hunk 12 above)
Comment (ospodaryk) [reply service/src/main/java/greencity/service/ubs/OrderBagService.java:104]: When I tested that method, it threw an error that "java.lang.UnsupportedOperationException: remove" so I had to replace it with this method

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java (hunk 20 above)
Comment (ospodaryk) [reply service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java:890]: Below we are using bagsForOrder so not sure if we should remove that

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java (hunk 22 above)
Comment (ospodaryk) [reply service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java:1193]: Throughout the program we use Long for price. Therefore, I am not sure how appropriate it is to change it

**Review by VadymMakitra — APPROVED** [review]

**Review by HelenSotnik — APPROVED** [review]

**Review by sergoliarnik — APPROVED** [review]
